serve/mlc_serve/model/tvm_model.py

```diff
@@ -67,36 +132,144 @@ def broadcast_from_worker_0(sess: di.Session, src, shape, dtype):
     return dst
 
 
+def load_tvm_weights(
+    param_path: Union[str, pathlib.Path], dev: tvm.runtime.Device
+) -> List[tvm.runtime.NDArray]:
+    """Load a set of weights
+
+    Loads a set of weights.  Assumes that the weights had previously
+    been saved using `tvmjs.dump_ndarray_cache`.
+
+    Parameters
+    ----------
+    param_path: Union[str, pathlib.Path]
+
+        The path to the saved parameters.
+
+    dev: tvm.runtime.Device
+
+        The device onto which the saved parameters should be loaded.
+
+    Returns
+    -------
+    params: List[tvm.runtime.NDArray]
+
+        The loaded parameters
+
+    """
+    param_path = pathlib.Path(param_path)
+
+    from tvm.contrib import tvmjs  # pylint: disable=import-outside-toplevel
+
+    params, metadata = tvmjs.load_ndarray_cache(param_path.as_posix(), dev)
+
+    try:
+        params_names = [
+            d["name"]
+            for d in json.loads(str(vm.module.get_function("_metadata")()))["params"]
+        ]
+    except:
+        params_names = [f"param_{i}" for i in range(metadata["ParamSize"])]
+
+    return [params[param_name] for param_name in param_names]
+
+
+def preprocess_weights(
+    param_path: Union[str, pathlib.Path],
+    executable_module: tvm.runtime.Module,
+    dev: tvm.runtime.Device,
+) -> List[tvm.runtime.NDArray]:
+    """Load and preprocess a set of model weights
+
+    Load and preprocess a set of model weights.
+
+    Parameters
+    ----------
+    param_path: Union[str, pathlib.Path]
+
+        The path to the model parameters parameters.
+
+    executable_module: tvm.runtime.Module
+
+        The compiled module, as returned from
+        `tvm.runtime.load_module`, which contains the functions
+        `transform_params` and `get_original_param_names`.
+
+    dev: tvm.runtime.Device
+
+        The device on which the pre-processed parameters should be
+        loaded, after pre-processing is complete.
+
+    Returns
+    -------
+    params: List[tvm.runtime.NDArray]
+
+        The loaded parameters
+
+    """
+    param_path = pathlib.Path(param_path)
+
+    safetensors = LazySafetensorDir(param_path)
```

Comment:
Is it safe to assume that safetensors are always available? If this reads from the HF model cards, I've seen the cases that this is not true.

Reply:
Conversion into safetensors format will be part of the asset ingestion step, so we should always have them available.  This is also on the fallback path if the `params` directory generated with `tvm.contrib.tvmjs` isn't available, so the normal workflow should still work as intended.

(Side-note: I like the safetensors format a lot more than the pytorch format, because we can read a single tensor without needing to unpack the entire file.  This avoids a large amount of caching that is needed when reading from pytorch files, as